### PR TITLE
[WIP] Log Entry Retry Backoff Mechanism

### PR DESCRIPTION
**Pull request by shreayp**
## Overview

Description:

This PR attempts to fix the following problem:

- The source has a set timeout period within the completable future for sent log entries. 
- In cases with larger batch sizes being sent to the sink, this timeout can repeatedly expire before an ack is received to complete the future.
- This prompts the source to resend the entires that timed out which will then keep accumulating on the sink, in some cases resulting in an OOM.

Why should this be merged: 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [ ] There are no TODOs left in the code
- [ ] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [ ] Change is covered by automated tests
- [ ] Public API has Javadoc

### CODE REVIEW

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java [hunk 1]
```diff
@@ -222,6 +228,12 @@ public void sendWithBuffering(List<LogReplicationEntryMsg> dataToSend, String me
     public LogReplicationEntryMsg resend() {
         LogReplicationEntryMsg ack = null;
         boolean force = false;
+
+        // Wait before retrying
+        if (isBackpressureActive.get()) {
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:233]: I think this check and wait must be done before sending, after processing the acks on L238

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java [hunk 2]
```diff
@@ -222,6 +228,12 @@ public void sendWithBuffering(List<LogReplicationEntryMsg> dataToSend, String me
     public LogReplicationEntryMsg resend() {
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:228]: This method gets invoked on Snapshot sync also.  For snapshot sync, the timeout reason will most likely be different.  We will have to decide if the backoff must be introduced there as well.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java [hunk 3]
```diff
@@ -219,11 +225,16 @@ public void sendWithBuffering(List<LogReplicationEntryMsg> dataToSend, String me
...
+    public LogReplicationEntryMsg resend(boolean isLogEntry) {
         LogReplicationEntryMsg ack = null;
         boolean force = false;
+
         try {
             ack = processAcks();
+
+            if (isLogEntry && isBackpressureActive.get()) {
+                waitBeforeRetry();
+            }
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:237]: I think a better back pressure mechanism would do the following:
1. consumer rejects message explicitly, returns busy response (as opposed to waiting for a timeout)
2. producer waits, then sends a probe message without the payload
3. if the probe returns not-busy, then the producer sends the payload, else keep probing

Comment (shreayp) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:237]: We had considered something similar by bounding the executor service that holds the RPCs on the sink, where any message beyond capacity is met will a null ack. I agree that having explicit messaging is a better approach but one objective of this fix was to minimally modify the current behavior, the introduction of additional RPCs might require more care.

**Review by hisundar — COMMENTED** [review]
The change looks good to me - is it possible to test this somehow?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java [hunk 4]
```diff
@@ -219,11 +236,22 @@ public void sendWithBuffering(List<LogReplicationEntryMsg> dataToSend, String me
...
+    public LogReplicationEntryMsg resend(boolean isLogEntry) {
         LogReplicationEntryMsg ack = null;
         boolean force = false;
+
         try {
             ack = processAcks();
+
+            if (isLogEntry && isBackpressureActive.get()) {
+                try {
+                    log.trace("Dropped ACKs from sink, resending log entry messages in some time...");
```
Comment (shama358) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:248]: we aren't dropping ACKs here, we are not getting it at all...so maybe reword this log statement?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java [hunk 5]
```diff
@@ -237,6 +265,12 @@ public LogReplicationEntryMsg resend() {
             }
         } catch (Exception e) {
             log.warn("Caught an exception while processing ACKs.", e);
+        } finally {
+            if (isLogEntry && force) {
+                activateBackpressure();
+            } else {
+                deactivateBackpressure();
```
Comment (shama358) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:272]: We would come here even if its a snapshot sync...can we restructure this block so we have a top level if statement checking for isLogEntry, and a child if-else block activating/deactivating the back pressure?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java [hunk 6]
```diff
@@ -260,6 +294,17 @@ public LogReplicationEntryMsg resend() {
         return ack;
     }
 
+    private void activateBackpressure() {
+        log.info("Log entry ACKs timing out on sink, enabling back pressure for retries.");
+        isBackpressureActive.set(true);
+    }
+
+    private void deactivateBackpressure() {
```
Comment (shama358) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:302]: We should also deactivate the back pressure when fsm transisitons out of log-entry sync.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java (hunk 5 above)
Comment (shreayp) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:272]: That behavior was intentional in the event that log entry was interrupted by a force sync then that snapshot sync would reset the back pressure when it calls resend, but deactivating when we transition out out log entry and making this explicit is a better approach so will take that up.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java (hunk 3 above)
Comment (pankti-m) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java]: yes, adding a 'busy' response on the receiver and probe request from the sender will require a more involved change which we had decided to avoid.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java [hunk 7]
```diff
@@ -219,11 +236,22 @@ public void sendWithBuffering(List<LogReplicationEntryMsg> dataToSend, String me
...
      */
-    public LogReplicationEntryMsg resend() {
+    public LogReplicationEntryMsg resend(boolean isLogEntry) {
         LogReplicationEntryMsg ack = null;
         boolean force = false;
+
         try {
             ack = processAcks();
+
+            if (isLogEntry && isBackpressureActive) {
```
Comment (shama358) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:246]: Are we going to come to this block incase we don't receive an ACK?
processAcks() waits for sometime and then times out..throwing a timeout exception in the process. So we won't be coming here right?

Comment (shama358) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:246]: Might have to move it to the for block below

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java [hunk 8]
```diff
@@ -219,11 +236,22 @@ public void sendWithBuffering(List<LogReplicationEntryMsg> dataToSend, String me
     /**
      * Resend the messages in the queue if they have timed out.
      */
-    public LogReplicationEntryMsg resend() {
+    public LogReplicationEntryMsg resend(boolean isLogEntry) {
         LogReplicationEntryMsg ack = null;
         boolean force = false;
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:241]: We can rename this variable to 'error' or something similar which is clearer.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SnapshotSender.java [hunk 9]
```diff
@@ -122,7 +122,7 @@ public void transmit(UUID snapshotSyncEventId, boolean forcedSnapshotSync) {
         if (Address.isAddress(baseSnapshotTimestamp)) {
             // Read and Send Batch Size messages, unless snapshot is completed before (endRead)
             // or snapshot sync is stopped
-            dataSenderBufferManager.resend();
+            dataSenderBufferManager.resend(false);
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SnapshotSender.java:125]: For future reference, can you add a comment stating why this is enabled for log entry sync only?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java (hunk 7 above)
Comment (shreayp) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:246]: sleeping for a smaller amount between each resend in the loop is probably better retry strategy, will take that up

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java [hunk 10]
```diff
@@ -42,6 +43,18 @@ public abstract class SenderBufferManager {
...
+     * Back of napkin math to derive sleep time (per message):
+     *  - Assume payload size of 15MB
+     *  - Queue holds ~50 messages before OOM risk
+     *  -> MAX_SLEEP_TIME_MS = 1 min to process 50 messages = 60000 / 50 = 1200 + buffer = 1350ms
+     *  -> INITIAL_SLEEP_TIME_MS = 20% of max time
+     *  -> SLEEP_TIME_INCREMENT_MS = 10% of max time
+     */
+    public static final long INITIAL_SLEEP_TIME_MS = 270;
+    public static final long MAX_SLEEP_TIME_MS = 1350;
+    public static final long SLEEP_TIME_INCREMENT_MS = 135;
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:56]: In the ideal case you only want one re-transmission on a request timeout. Since this workflow is not latency sensitive, I suggest starting with a higher timeout (i.e., seconds) to reduce the probability of retrying more than once.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java [hunk 11]
```diff
@@ -219,29 +245,44 @@ public void sendWithBuffering(List<LogReplicationEntryMsg> dataToSend, String me
...
+            }
         }
 
         for (int i = 0; i < pendingMessages.getSize(); i++) {
             LogReplicationPendingEntry entry = pendingMessages.getPendingEntries().get(i);
-            if (entry.timeout(msgTimer) || force) {
+            if (entry.timeout(msgTimer) || error) {
+                // Upon an ACK failing due to a timeout, we add a linearly increasing wait between retries as
+                // a backpressure mechanism.
+                if (isLogEntry && isBackpressureActive) {
```
Comment (shama358) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:282]: rethinking the placement: I think currently the number of inflight msgs is 1, so this change works
But to be future proof, should we move this logic just before the for loop? 
I am thinking of this scenario: Say we allow 5 msgs be in flight, should we wait for sometime and send all the 5 msgs without a delay, or do we want a delay before sending each msg? 
I am inclining towards the former apprach as the latter will increase the latency even when not necessary, but would like to know yours/reviewers thoughts.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java [hunk 12]
```diff
@@ -42,6 +43,18 @@ public abstract class SenderBufferManager {
      */
     public static final String config_file = "/config/corfu/corfu_replication_config.properties";
 
+    /*
+     * Back of napkin math to derive sleep time (per message):
+     *  - Assume payload size of 15MB
+     *  - Queue holds ~50 messages before OOM risk
+     *  -> MAX_SLEEP_TIME_MS = 5 min to process 50 messages = 300000 / 50 = 6000 + buffer = 7000ms
```
[the remark this answers is not in the document]
Comment (shreayp) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:50]: From @pankti-m's initial triage, 5 minutes was roughly the amount of time required to apply all the transaction in the entry that was being repeatedly sent. Nothing concrete, these values can/will be adjusted after findings from scale testing.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java (hunk 11 above)
Comment (shreayp) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:282]: In the case that we saw there were 2 entires, asymmetric in size with one being around 20mb and the other 8mb. Presumably the 20mb one resulted in a long running apply leading to the resends. If we go with the approach to send them in bursts, a flight of messages is likely to be resent in the event that one of them does not complete within that 5sec allotted by the enclosing future, since all the entries will be sent at roughly the same time. My thinking was that by distributing the wait across entries, you now have the added benefit of the futures expiring sequentially.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java [hunk 13]
```diff
@@ -111,11 +128,22 @@ public SenderBufferManager(DataSender dataSender) {
...
 
         readConfig();
         pendingMessages = new SenderPendingMessageQueue(maxBufferSize);
         pendingCompletableFutureForAcks = new HashMap<>();
         this.dataSender = dataSender;
+
+        // Back of napkin math to derive sleep time (per message):
+        // Assume payload size of 15MB
+        // Queue holds ~50 messages before OOM risk
+        // maxSleepTime (default) = 5min to process 50 messages = 300000 / 50 = 6000 + buffer = 7000ms
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:141]: is 5mins an arbitrary number for the max wait across 50 messages?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java (hunk 11 above)
Comment (pankti-m) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:285]: I was under the impression that we decided with the approach which adds a backoff before each message.

I am open to the other approach also as this change does have tuning which will be done incrementally.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java [hunk 14]
```diff
@@ -313,4 +373,43 @@ private CompletableFuture<LogReplicationEntryMsg> timeEntrySend(Timer.Sample sam
...
+            this.sleepTimeIncrement = sleepTimeIncrement;
+            this.currentSleepTime = initialSleepTime;
+        }
+
+        public void backoff() {
+            try {
+                log.trace("ACKs timing out on sink, resending messages in {} milliseconds...", currentSleepTime);
+                TimeUnit.MILLISECONDS.sleep(currentSleepTime);
+            } catch (Exception e) {
+                Thread.currentThread().interrupt();
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:402]: why do we interrupt the current thread here?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java (hunk 11 above)
Comment (shreayp) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:285]: Yes, the current implementation has a wait before each message

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java (hunk 13 above)
Comment (shreayp) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:141]: Not completely arbitrary, it was the time needed to apply the transactions the 20mb entry from your initial triage, but also not concrete which is why I've went ahead and made that parameter tunable.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java (hunk 14 above)
Comment (shreayp) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:402]: Had this here from when I had the wait set to some order of minuets, so the interrupted status is preserved and the next call to sleep exits without waiting until the interrupt flag is reset which would come when `processAcks()` hits a InterruptedException on the next blocking call and clears it implicitly. Can remove this for now since the logic has changed since and reinterrupting the thread doesn't appear to have any benefits

File: infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java [hunk 15]
```diff
@@ -244,6 +246,34 @@ public ExecutorService getExecutorService(int threadCount, @Nonnull ThreadFactor
...
+     * @param threadCount    The number of threads to use in the pool
+     * @param threadFactory  The underlying thread factory
+     * @param queueCapacity  Max capacity of underlying linked blocking queue
+     * @return The newly created ExecutorService
+     */
+    public ExecutorService getBoundedExecutorService(int threadCount, @Nonnull ThreadFactory threadFactory, int queueCapacity) {
+        return new ThreadPoolExecutor(threadCount, threadCount, 0L, TimeUnit.MILLISECONDS,
+                new LinkedBlockingQueue<>(queueCapacity), threadFactory, new AbortPolicy());
+    }
+
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java:276]: This seems like its a little verbose and no generic for reuse. I think its cleaner to just get the number of pending requests in `processRequest` with `((ThreadPoolExecutor)executor).getQueue().size()`

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientRouter.java [hunk 16]
```diff
@@ -136,7 +134,7 @@ public LogReplicationClientRouter(LogReplicationRuntimeParameters parameters,
         this.remoteClusterDescriptor = parameters.getRemoteClusterDescriptor();
         this.remoteClusterId = remoteClusterDescriptor.getClusterId();
         this.parameters = parameters;
-        this.timeoutResponse = parameters.getRequestTimeout().toMillis();
+        TIMEOUT_RESPONSE = parameters.getRequestTimeout().toMillis();
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientRouter.java:137]: I think we need to keep the original timeout. Instead if you want to set larger timeouts for specific requests you can overload ` sendRequestAndGetCompletable`  to pass a timeout value

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java [hunk 17]
```diff
@@ -260,6 +309,28 @@ public LogReplicationEntryMsg resend() {
         return ack;
     }
 
+    private void activateBackpressure() {
+        if (!isBackpressureActive) {
+            log.info("Log entry ACKs timing out on sink, allocating additional time for retries.");
+            isBackpressureActive = true;
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:315]: Why do you need to switch this on/off?

Comment (shreayp) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:315]: I saw some getters/setters being used for the timeoutResponse within the failure detector, shouldn't be applicable for LR's router implementation but to keep the value consistent when we want to add a retry the on/off just updates the static field and resets to the original value it has cached, also related to your above comment. But if there's no need for that field to be real time accurate, overloading `sendRequestAndGetCompletable` is a cleaner approach

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/LogReplicationConfig.java [hunk 18]
```diff
@@ -48,7 +48,7 @@ public class LogReplicationConfig {
     public static final int MAX_CACHE_NUM_ENTRIES = 200;
 
     // Percentage of log data per log replication message
-    public static final int DATA_FRACTION_PER_MSG = 90;
+    public static final long DATA_FRACTION_PER_MSG = 90L;
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/LogReplicationConfig.java:51]: I suppose changing the data-type here will not have any impact as the final variable is int.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/LogReplicationConfig.java [hunk 19]
```diff
@@ -39,7 +39,7 @@ public class LogReplicationConfig {
 
     // Max message size supported by protocol buffers is 64MB. Log Replication uses this limit as the default message
     // size to batch and send data across over to the other side.
-    public static final int MAX_DATA_MSG_SIZE_SUPPORTED = (64 << 20);
+    public static final int MAX_DATA_MSG_SIZE_SUPPORTED = (24 << 20);
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/LogReplicationConfig.java:42]: Shouldnt this value be 25 MB?  Can you also update the comment to reflect the change and how 25 MB was decided (default max tx size supported for Corfu tx)

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/LogReplicationConfig.java (hunk 18 above)
Comment (shreayp) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/LogReplicationConfig.java:51]: When it is used to compute the maxDataSizePerMsg `((maxMsgSize * DATA_FRACTION_PER_MSG) / 100L)` the multiplication before the division will happen with long division avoiding the overflow

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientRouter.java [hunk 20]
```diff
@@ -181,6 +181,22 @@ public IClientRouter addClient(IClient client) {
...
+    /**
+     * Send a request message and get a completable future to be fulfilled by the reply within
+     * supplied timeout response time.
+     *
+     * @param payload
+     * @param <T> The type of completable to return.
+     * @return A completable future which will be fulfilled by the reply,
+     * or a timeout in the case there is no response within the allotted time.
+     */
+    public <T> CompletableFuture<T> sendRequestAndGetCompletable(
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientRouter.java:196]: This will send the message immediately but wait for a longer time for the response.  Instead, hadnt we decided to wait for this time before re-sending?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java [hunk 21]
```diff
@@ -260,6 +302,26 @@ public LogReplicationEntryMsg resend() {
         return ack;
     }
 
+    private void activateBackpressure() {
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:305]: these 2 methods are for metrics only.  Can you rename them to reflect the same?  Something like logBackPressureActivationMetrics() or something similar.

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/LogReplicationConfig.java (hunk 18 above)
Comment (pankti-m) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/LogReplicationConfig.java:51]: but is it required now with the max size being 25MB?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientRouter.java (hunk 20 above)
Comment (shreayp) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientRouter.java:196]: I think we had found that either was suitable, however doing it this way we avoid introducing a sleep in our sender while getting the same effect of the added back pressure

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java (hunk 21 above)
Comment (shreayp) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SenderBufferManager.java:305]: In addition to the metrics they also flip that boolean field to indicate when the timeouts should continue to be extended

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/LogReplicationConfig.java (hunk 18 above)
Comment (shreayp) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/LogReplicationConfig.java:51]: yes even down to 25MB the value still comes right outside the int range

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/LogReplicationConfig.java [hunk 22]
```diff
@@ -37,9 +37,9 @@ public class LogReplicationConfig {
     // Default value for the max number of entries applied in a single transaction on Sink during snapshot sync
     public static final int DEFAULT_MAX_SNAPSHOT_ENTRIES_APPLIED = 50;
 
-    // Max message size supported by protocol buffers is 64MB. Log Replication uses this limit as the default message
+    // Max message size supported by protocol buffers is 25MB. Log Replication uses this limit as the default message
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/LogReplicationConfig.java:40]: This is incorrect.  Max message size supported by protocol buffers is not 25 MB.   This value is the default Corfu transaction size.  Can you please update the comment?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/LogReplicationConfig.java [hunk 23]
```diff
@@ -37,9 +37,9 @@ public class LogReplicationConfig {
...
     // size to batch and send data across over to the other side.
-    public static final int MAX_DATA_MSG_SIZE_SUPPORTED = (64 << 20);
+    public static final int MAX_DATA_MSG_SIZE_SUPPORTED = 25 * 1_024 * 1_024;
```
Comment (pankti-m) [inline infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/LogReplicationConfig.java:42]: is this a typo?

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientRouter.java (hunk 20 above)
Comment (pankti-m) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientRouter.java:196]: hmm... I may be mistaken but I dont remember deciding this.  Backoff essentially means waiting before retrying...

File: infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/LogReplicationConfig.java (hunk 23 above)
Comment (shreayp) [reply infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/LogReplicationConfig.java:42]: No this is intentional, same thing as bit shifting left by 20 but felt this was more indicative of the size in bytes
